File: 한민선/240528/2504. 괄호의 값/괄호의 값.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        Stack<Character> stack = new Stack<>();

        // (,[ 괄호일 때 *2, *3
        // ),] 괄호일 때 /2, /3 -> 괄호가 닫혔을 때 곱셈을 초기화해줌
        // 괄호가 닫혔을 때(곱셈 계산이 끝남) 곱셈한 결과를 이전 괄호 값에 더해줌

        // 곱셈 계산 저장할 변수
        int imsy = 1;
        int answer = 0;

        char[] input = br.readLine().toCharArray();
        for (int i = 0; i < input.length; i++) {
            switch (input[i]){
                case '(':
                    stack.push(input[i]);
                    imsy *= 2;
                    break;
                case ')':
                    if (!stack.empty() && stack.peek() == '('){
                        if (input[i-1] == '(') {
                            // input의 괄호가 () 이 형태일 때 최종 계산 => input의 이전 값 확인
                            answer += imsy;
                            imsy /= 2;
                            stack.pop();
                        } else{
                            imsy /= 2;
                            stack.pop();
                        }
                    }else{
                        System.out.println("0");
                        return;
                    }
                    break;
                case '[':
                    stack.push(input[i]);
                    imsy *= 3;
                    break;
                case ']':
                     if (!stack.empty() && stack.peek() == '['){
                         if (input[i-1] == '[') {
                             answer += imsy;
                             imsy /= 3;
                             stack.pop();
                         } else{
                             imsy /= 3;
                             stack.pop();
                         }
                     }else{
                         System.out.println("0");
                         return;
                     }
                    break;
            }
        }
        if(imsy != 1){
            answer = 0;
        }
        bw.write(answer + "");
        bw.flush();
        bw.close();
    }
}

File: 이하영/240528/2504. 괄호의 값/괄호의 값.py
# 1. 괄호 리스트로 받아옴
# 2. 반복문을 돌리다가
# 3. (,[ 가 나오면 stack에 push, 괄호에 맞는 값을 변수에 곱해줌
# 4. ), ] 가 나오면 
# 4-1. 직전에 들어있는 괄호가 짝이 맞는 괄호 인지 확인 후 맞으면 덧셈 연산 
# 4-2. 짝이 안맞으면 0 출력 후 종료
# 4-3. 스택 pop 후(괄호의 짝을 찾았으니까), 해당 값에 맞는 나누기 연산(열린 만큼 곱해준 변수를 다음 연산을 위해 1로 만들어주는 작업)


import sys

brackets = sys.stdin.readline()
stack = []
answer = 1 #곱하기만 한 값을 저장
result = 0 #더하기까지 한 결과를 저장

for i in range (len(brackets)):
    if brackets[i] == "(" :
        stack.append(brackets[i])
        answer *= 2
    elif brackets[i] == "[" :
        stack.append(brackets[i])
        answer *= 3
    elif brackets[i] == ")":
        if brackets[i-1] == "(":
            result += answer
        elif not stack or stack[-1] != "(":
            result = 0
            break
        stack.pop()
        answer //= 2
    elif brackets[i] == "]":
        if brackets[i-1] == "[":
            result += answer
        elif not stack or stack[-1] != "[":
            result = 0
            break
        stack.pop()
        answer //= 3

if stack:
    print(0) #짝이 맞지 않아서 남은 경우
else:
    print(result)